### PR TITLE
Prepare mapping API for D3D11

### PR DESCRIPTION
**Pull request by Bastacyclop**
Following #1098.
@snuk182 @kvark

### CODE REVIEW

File: src/backend/gl/src/factory.rs [hunk 1]
```diff
@@ -216,13 +216,15 @@ pub enum MappingKind {
...
-#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+#[derive(Debug, Eq, Hash, PartialEq)]
 #[allow(missing_copy_implementations)]
 pub struct MappingGate {
     pub kind: MappingKind,
     pub pointer: *mut ::std::os::raw::c_void,
     pub target: gl::types::GLenum,
-    pub is_mapped: bool,
+    // TODO: put those inside the kind enum or something ?
+    pub status: mapping::Status<R>, // Persistent Only
```
Comment (kvark) [inline src/backend/gl/src/factory.rs]: yes, please, let's use `enum`

File: src/core/src/lib.rs [hunk 2]
```diff
@@ -182,7 +182,7 @@ pub trait Resources:          Clone + Hash + Debug + Eq + PartialEq + Any {
     type DepthStencilView:    Clone + Hash + Debug + Eq + PartialEq + Any + Send + Sync;
     type Sampler:             Clone + Hash + Debug + Eq + PartialEq + Any + Send + Sync + Copy;
     type Fence:               Clone + Hash + Debug + Eq + PartialEq + Any + Send + Sync;
-    type Mapping:             Clone + Hash + Debug + Eq + PartialEq + Any + Send + Sync + mapping::Gate<Self>;
+    type Mapping:             Hash + Debug + Eq + PartialEq + Any + Send + Sync + mapping::Gate<Self>;
```
Comment (kvark) [inline src/core/src/lib.rs]: I hope this does not prevent automatic deriving of `Clone` for all the handles parametrized by `R`

Comment (Bastacyclop) [inline src/core/src/lib.rs]: Well `R` is still `Clone` right, I don't think an associated type will change that ?

Comment (kvark) [inline src/core/src/lib.rs]: ok, good

**Comment by sectopod** [comment]
Instancing example fails on D3D11 because of the `unimplemented!` arm in `map_usage`.
Even if I force it to be a dynamic buffer, D3D11 doesn't like it. There are no resources in D3D11 that are both readable and writable by the CPU, and accessible by the GPU. A dynamic resource is only writable on CPU, and staging resources can't be used for anything on GPU other than transfer operations.

One solution I could propose is:
  - change the instancing example (and others) to use Write-only mapping
  - fix `map_usage` to use the dynamic buffer type, but also make sure Read is not requested there on D3D11 only, returning an error

**Comment by snuk182** [comment]
@sectopod fixing `map_usage` itself would not help, as there is a ban on using mappings for anything except Persistent ( @Bastacyclop will rename it to Mappable for clarity ), so either the example should be rewritten from `map_buffer` to `update_buffer` or mapping for DX11 should be eventually implemented (working on it, but not as fast as I'd want to). And yes, `RW` map access paradigm is indeed not portable, so should be rethinked...

**Comment by Bastacyclop** [comment]
I am in favor of rewriting the example to use write-only mapping and returning something like `mapping::Error::AccessNotSupported(access)` when appropriate.

**Comment by kvark** [comment]
ok, then, looks like a progress :)
@homu r+

**Pull request merged**